File: 101-200/106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
class Solution {
    unordered_map<int, int> hash;
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        for (int i = 0; i < inorder.size(); ++i) {
            hash[inorder[i]] = i;
        }
        return helper(inorder, postorder, 0, inorder.size()-1, 0, postorder.size()-1);
    }

    TreeNode* helper(vector<int>& inorder, vector<int>& postorder, int inStart, int inEnd, int postStart, int postEnd) {
        if (postStart > postEnd) return nullptr;
        int nodeVal = postorder[postEnd];
        TreeNode* root = new TreeNode(nodeVal);
        int index = hash[nodeVal];
        int leftLen = index - inStart;
        root->left = helper(inorder, postorder, inStart, index-1, postStart, postStart+leftLen-1);
        root->right = helper(inorder, postorder, index+1, inEnd, postStart+leftLen, postEnd-1);
        return root;
    }
};

File: python version/106-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:

        def helper(in_start, in_end, post_end):
            if in_start>in_end: return None
            val = postorder[post_end]
            node = TreeNode(val)
            in_pos = inorder.index(val)
            right_len = in_end-in_pos
            node.left = helper(in_start, in_pos-1, post_end-right_len-1)
            node.right = helper(in_pos+1, in_end, post_end-1)
            return node

        return helper(0, len(inorder)-1, len(postorder)-1)